clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/ElectionPolicyBuilder.java

```diff
@@ -0,0 +1,94 @@
+import org.jboss.as.clustering.dmr.ModelNodes;
+import org.jboss.as.controller.ExpressionResolver;
+import org.jboss.as.controller.OperationFailedException;
+import org.jboss.as.network.OutboundSocketBinding;
+import org.jboss.dmr.ModelNode;
+import org.jboss.msc.service.ServiceBuilder;
+import org.jboss.msc.service.ServiceController;
+import org.jboss.msc.service.ServiceTarget;
+import org.jboss.msc.service.ValueService;
+import org.jboss.msc.value.Value;
+import org.wildfly.clustering.service.Builder;
+import org.wildfly.clustering.service.Dependency;
+import org.wildfly.clustering.service.InjectedValueDependency;
+import org.wildfly.clustering.singleton.SingletonElectionPolicy;
+import org.wildfly.clustering.singleton.election.NamePreference;
+import org.wildfly.clustering.singleton.election.Preference;
+import org.wildfly.clustering.singleton.election.PreferredSingletonElectionPolicy;
+import org.wildfly.extension.clustering.singleton.election.OutboundSocketBindingPreference;
+
+/**
+ * Builds a service that provides an election policy.
+ * @author Paul Ferraro
+ */
+public abstract class ElectionPolicyBuilder extends ElectionPolicyServiceNameProvider implements ResourceServiceBuilder<SingletonElectionPolicy>, Value<SingletonElectionPolicy> {
+
+    private final List<Preference> preferences = new CopyOnWriteArrayList<>();
+    private final List<Dependency> dependencies = new CopyOnWriteArrayList<>();
+
+    protected ElectionPolicyBuilder(String name) {
+        super(name);
+    }
+
+    @Override
+    public ServiceBuilder<SingletonElectionPolicy> build(ServiceTarget target) {
+        final List<Preference> preferences = this.preferences;
+        Value<SingletonElectionPolicy> value = new Value<SingletonElectionPolicy>() {
+            @Override
+            public SingletonElectionPolicy getValue() {
+                SingletonElectionPolicy policy = ElectionPolicyBuilder.this.getValue();
+                return preferences.isEmpty() ? policy : new PreferredSingletonElectionPolicy(policy, preferences);
+            }
+        };
+        ServiceBuilder<SingletonElectionPolicy> builder = target.addService(this.getServiceName(), new ValueService<>(value)).setInitialMode(ServiceController.Mode.ON_DEMAND);
+        for (Dependency dependency : this.dependencies) {
+            dependency.register(builder);
+        }
+        return builder;
+    }
+
+    @Override
+    public Builder<SingletonElectionPolicy> configure(ExpressionResolver resolver, ModelNode model) throws OperationFailedException {
+        this.preferences.clear();
+        this.dependencies.clear();
+        for (String bindingName : ModelNodes.asStringList(Attribute.SOCKET_BINDING_PREFERENCES.getDefinition().resolveModelAttribute(resolver, model))) {
+            InjectedValueDependency<OutboundSocketBinding> binding = new InjectedValueDependency<>(Capability.SOCKET_BINDING_PREFERENCE.getDefinition().getCapabilityServiceName(bindingName), OutboundSocketBinding.class);
```

Comment:
I really think the `OperationContext` should be used here so the capability can correctly be resolved. The context can be important especially after https://github.com/wildfly/wildfly-core/pull/917 is merged.